bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/googletv/GoogleTVConnectionManager.java

```diff
@@ -0,0 +1,1383 @@
+                        if (!isLoggedIn) {
+                            if (command.toString().equals("REQUEST")) {
+                                sendCommand(new GoogleTVCommand(
+                                        GoogleTVRequest.encodeMessage(GoogleTVRequest.pinRequest(command.toString()))));
+                            } else if (shimServerChain != null) {
+                                this.pinHash = GoogleTVUtils.validatePIN(command.toString(), androidtvPKI.getCert(),
+                                        shimServerChain[0]);
+                                sendCommand(new GoogleTVCommand(
+                                        GoogleTVRequest.encodeMessage(GoogleTVRequest.pinRequest(this.pinHash))));
+                            }
+                        }
+                    } else if ((config.mode.equals(PIN_MODE)) && (config.shim)) {
+                        if ((shimClientChain != null) && (shimServerChain != null) && (shimClientLocalChain != null)) {
+                            this.pinHash = GoogleTVUtils.validatePIN(command.toString(), androidtvPKI.getCert(),
+                                    shimServerChain[0]);
+                            this.shimPinHash = GoogleTVUtils.validatePIN(command.toString(), shimClientChain[0],
+                                    shimClientLocalChain[0]);
+                        }
+                    }
+                } catch (CertificateException e) {
+                    logger.trace("PIN CertificateException", e);
+                }
+            }
+        } else if (CHANNEL_POWER.equals(channelUID.getId())) {
+            if (command instanceof OnOffType) {
+                if ((power && command.equals(OnOffType.OFF)) || (!power && command.equals(OnOffType.ON))) {
+                    sendCommand(new GoogleTVCommand(GoogleTVRequest.encodeMessage("5204081a1003")));
+                }
+            } else if (command instanceof StringType) {
+                if ((power && command.toString().equals("OFF")) || (!power && command.toString().equals("ON"))) {
+                    sendCommand(new GoogleTVCommand(GoogleTVRequest.encodeMessage("5204081a1003")));
+                }
+            }
+        } else if (CHANNEL_MUTE.equals(channelUID.getId())) {
+            if (command instanceof OnOffType) {
+                if ((volMute && command.equals(OnOffType.OFF)) || (!volMute && command.equals(OnOffType.ON))) {
+                    sendCommand(new GoogleTVCommand(GoogleTVRequest.encodeMessage("5204085b1003")));
+                }
+            }
+        } else if (CHANNEL_DEBUG.equals(channelUID.getId())) {
+            if (command instanceof StringType) {
+                if (command.toString().startsWith("RAW", 9)) {
+                    String newCommand = command.toString().substring(13);
+                    String message = GoogleTVRequest.encodeMessage(newCommand);
+                    logger.trace("Raw Message Decodes as: {}", GoogleTVRequest.decodeMessage(message));
```

Comment:
Decoding is probably non-trivial so it would be better to make this conditional.